napari/layers/image/_slice.py

```diff
@@ -255,14 +264,53 @@ def _call_multi_scale(self) -> _ImageSliceResponse:
             request_id=self.id,
         )
 
-    def _slice_indices_at_level(self, level: int) -> np.ndarray:
-        indices = np.array(self.indices)
-        axes = self.dims.not_displayed
-        ds_indices = indices[axes] / self.downsample_factors[level][axes]
-        ds_indices = np.round(ds_indices.astype(float)).astype(int)
-        ds_indices = np.clip(ds_indices, 0, self.level_shapes[level][axes] - 1)
-        indices[axes] = ds_indices
-        return indices
+    def _thick_slice_at_level(self, level: int) -> _ThickNDSlice:
+        """
+        Get the data_slice rescaled for a specific level.
+        """
+        slice_arr = self.data_slice.as_array()
+        # downsample based on level
+        slice_arr /= self.downsample_factors[level]
+        slice_arr[0] = np.clip(slice_arr[0], 0, self.level_shapes[level] - 1)
+        return _ThickNDSlice.from_array(slice_arr)
+
+    def _project_thick_slice(self, data, data_slice):
+        """
+        Slice the given data with the given data slice and project the extra dims.
+        """
+        slices = [
+            slice(None) if np.isnan(p) else int(np.round(p))
+            for p in data_slice.point
+        ]
+
+        if self.projection_mode == 'none':
+            # early return with only the dims point being used
+            return np.asarray(data[tuple(slices)])
```

Comment:
I think the main reason for `asarray` here is because `Image.data` only needs to implement the layer data protocol whereas vispy wants a numpy array. In the case that `data` needs to be materialized/evaluated (e.g. a dask array), then I think we want to do it here because at least we have a chance of doing it off the main thread (e.g. when async slicing is enabled).

We could special case for dask, but I don't see the benefit because the [numpy docs](https://numpy.org/doc/stable/reference/generated/numpy.asarray.html#numpy-asarray) claim that `asarray` only makes a copy if it needs to and in this case it shouldn't if `data` is a numpy array because we're not changing the dtype or order. Is that understanding wrong? Should we use `np.array(..., copy=False)` instead?

The only other potential copy I can see is indexing `data` (because of [numpy advanced indexing](https://numpy.org/doc/stable/user/basics.indexing.html#advanced-indexing)) , but I think the index is always a `tuple[int | slice, ...]` so that shouldn't happen.